src/transformStateWithClones.js

```diff
@@ -7,7 +7,26 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const history = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    let newState = { ...currentState };
+
+    if (action.type === 'addProperties') {
+      Object.assign(newState, action.extraData);
+    } else if (action.type === 'removeProperties') {
+      action.keysToRemove.forEach((key) => {
+        delete newState[key];
+      });
+    } else if (action.type === 'clear') {
+      newState = {};
+    }
+    history.push({ ...newState });
```

Comment:
Pushing a copy of the object to the `history` array should be done once after the current action is processed and outside of the `switch` block. This will adhere to the DRY principle and reduce redundancy.